### ISSUE TITLE
fta-cli: limiting output does not show most complex files, and skips checking the score entirely

### ISSUE THREAD

**Issue by gabriel-v** [sgb-io/fta#46]
The `output_limit` config for fta-cli doesn't do what I expect it to do.

It halts processing after that number of files is reached, instead of running on all files and showing first lines of sorted output.

Here is my output with a limit:

```
┌──────────────────────────┬────────────┬─────────────────────────────┬────────────┐
│ File                     ┆ Num. lines ┆ FTA Score (Lower is better) ┆ Assessment │
╞══════════════════════════╪════════════╪═════════════════════════════╪════════════╡
│ next.config.js           ┆ 192        ┆ 18.03                       ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ styles/error.ts          ┆ 56         ┆ 14.94                       ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ generate-nginx-config.js ┆ 24         ┆ 13.63                       ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ postinstall-fixes.js     ┆ 13         ┆ 11.57                       ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/index.tsx          ┆ 16         ┆ 11.20                       ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/maps.tsx           ┆ 13         ┆ 8.75                        ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/insights.tsx       ┆ 7          ┆ 7.43                        ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/uploads.tsx        ┆ 7          ┆ 7.43                        ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/404.tsx            ┆ 5          ┆ 6.82                        ┆ OK         │
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┤
│ pages/terms.tsx          ┆ 5          ┆ 6.82                        ┆ OK         │
└──────────────────────────┴────────────┴─────────────────────────────┴────────────┘
10 files analyzed in 0.0139s.

```

But here is the full output with no output limit:

```
┌────────────────────────────────────────────────────┬────────────┬─────────────────────────────┬───────────────────┐                                                                                                                                         
│ File                                               ┆ Num. lines ┆ FTA Score (Lower is better) ┆ Assessment        │                                                                                                                                         
╞════════════════════════════════════════════════════╪════════════╪═════════════════════════════╪═══════════════════╡                                                                                                                                         
│ src/backend/buildSearchQuery.ts                    ┆ 438        ┆ 72.85                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/components/document/Document.tsx               ┆ 376        ┆ 72.25                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ ...nents/document/SubTabs/components/Tags/Tags.tsx ┆ 360        ┆ 69.70                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ ...h/results/resultsList/ResultItem/ResultItem.tsx ┆ 354        ┆ 69.65                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/stores/DocumentStore.tsx                       ┆ 292        ┆ 67.88                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/utils/utils.tsx                                ┆ 271        ┆ 67.59                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/components/pdf-viewer/Document/Document.tsx    ┆ 265        ┆ 66.67                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/components/pdf-viewer/Toolbar/Toolbar.tsx      ┆ 252        ┆ 65.22                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ ...mponents/search/filters/Histogram/Histogram.tsx ┆ 239        ┆ 65.00                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/stores/TagsStore.ts                            ┆ 200        ┆ 64.10                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/stores/search/FiltersStore.ts                  ┆ 219        ┆ 63.84                       ┆ Needs improvement │                                                                                                                                         
├╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┼╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌╌┤                                                                                                                                         
│ src/components/search/Search.tsx                   ┆ 203        ┆ 63.79                       ┆ Needs improvement │   


.... 
```

If I combine this with a max score `score_cap`, the `output_limit` prevents throwing error above that score.

You can try it out on our codebase: https://github.com/liquidinvestigations/hoover-ui

**Comment by sgb-io** [comment]
Thanks for your interest in the project @gabriel-v, I hope it's useful :)

I think you are right and this is a bug - see [here](https://github.com/sgb-io/fta/blob/dce0a93e1a69eef9a2616c8ae07cf99dd9953960/crates/fta/src/lib.rs#L138), thanks for reporting it. The analysis should be carried out regardless of `output_limit`; `output_limit` should only apply to the actual output.

I think a new option might also be appropriate like `max_files_to_analyze` since I've used `output_limit` to avoid crashing when scanning huge directories

**Comment by sgb-io** [comment]
For any other observers, in case it's not obvious, a workaround is to simply set `output_limit` to a high number to ensure the whole project is scanned

**Comment by sgb-io** [comment]
This is resolved in v1.0.0
